server/buildbuddy_server/buildbuddy_server.go

```diff
@@ -1138,38 +1138,56 @@ func (s *BuildBuddyServer) getAnyAPIKeyForInvocation(ctx context.Context, invoca
 // them up from our cache servers using the bytestream API or pulling them
 // from blobstore.
 func (s *BuildBuddyServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	params := r.URL.Query()
+	s.serveUsingParams(w, r, r.URL.Query())
+}
 
+func (s *BuildBuddyServer) serveUsingParams(w http.ResponseWriter, r *http.Request, params url.Values) {
+	var code int
+	var err error
 	if params.Get("artifact") != "" {
-		s.serveArtifact(r.Context(), w, params)
-		return
-	}
-	if params.Get("bytestream_url") != "" {
+		code, err = s.serveArtifact(r.Context(), w, params)
+	} else if params.Get("bytestream_url") != "" {
 		// bytestream request
-		s.serveBytestream(r.Context(), w, params)
+		code, err = s.serveBytestream(r.Context(), w, params)
+	} else {
+		http.Error(w, `One of "artifact" or "bytestream_url" query param is required`, http.StatusBadRequest)
 		return
 	}
-	http.Error(w, `One of "artifact" or "bytestream_url" query param is required`, http.StatusBadRequest)
+	if err != nil {
+		if code == http.StatusNotFound {
+			// Attempt to fall back gracefully if the specified file is not found and
+			// a fallback is specified.
+			if fallback, parseErr := url.Parse(params.Get("with_fallback")); parseErr == nil && fallback.RawQuery != "" {
+				q := fallback.Query()
+				q.Del("with_fallback") // Only fall back once.
+				s.serveUsingParams(w, r, q)
+				return
+			} else if parseErr != nil {
+				log.Infof("Parse Error: %s", parseErr)
+			} else {
+				log.Warningf("Fallback without query is invalid: %v", fallback)
+			}
```

Comment:
nit: we can return these as HTTP client errors to avoid adding log noise

Reply:
Done.